File: 206-reverse-linked-list/206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode listhead = null;
    public ListNode reverseList(ListNode head) {
        helper(null, head);
        return listhead;
    }

    private void helper(ListNode prev, ListNode curr) {
        if(curr == null) {
            listhead = prev;
        }
        else {
            helper(curr, curr.next);
            curr.next = prev;
        }
    }
}

File: 206-reverse-linked-list/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        pre = None
        curr = head
        
        while curr:
            nxt = curr.next
            curr.next = pre
            pre = curr
            curr = nxt
        
        return pre
        